Application Server session recovery improvements

adriansmares opened the pull request:
<!--
Thanks for submitting a pull request. Please fill the template below,
otherwise we will not be able to process this pull request.
-->

#### Summary
<!--
A short summary, referencing related issues:
Closes #0000, References #0000, etc.
-->

Closes #4067 

#### Changes
<!-- What are the changes made in this pull request? -->

- Broke down `as.downlinkQueueOp` as it was huge and did too many things at once:
  -  Observation logic has been moved to `registerReceiveDownlinks` / `as.registerDropDownlinks` / `as.registerForwardDownlinks`
  - The retry loop (that basically syncs the session between AS and NS) is now part of `as.attemptDownlinkQueueOp`
  - The downlink encode + encryption is now part of `as.encodeAndEncryptDownlinks`
- Flattened `as.handleJoinAccept` near the 'invalidate downlinks', as the nesting made it very hard to see what was happening.
- As part of the JoinAccept handling, the invalidated downlinks may be decrypted using what is called a 'ternary session'. A ternary session is simply the old `dev.PendingSession`. This occurs when the device attempted to join the network twice, but didn't confirm the session via an uplink.
- `as.matchSession` no longer recalculates the queue when it doesn't recognize the session (i.e. it's not `dev.Session`, nor `dev.PendingSession`). The assumption here is that the Network Server has thrown away the 'old' queue for what is probably a long time, as session confirmation swaps the pending session queue with the current queue. If the AS is in the position in which it never actually saw this session before, there is no way the queue in the Network Server is not empty (as there is no other entity that could have touched the session queue, or the pending session queue).
- Downlink queue invalidations are now handled via pushes, instead of replaces. The Network Server 'evicts' this downlink queue upon invalidation, so we have no reason to actually replace the queue. This protects us from cases during which we wrongfully replace valid items inserted by an application between the invalidation and the handling.
- Downlink nacks are now handled via pushes, in order to avoid replacing a valid queue with an old queue. I understand that this behavior is changed from what we did before, but I bring the following arguments:
  - The NS is already allowed to schedule the next downlinks in the queue - we're just racing with it if we attempt to put this item at the front of the queue.
  - When https://github.com/TheThingsNetwork/lorawan-stack/issues/3642 will be done, the NS will automatically retry the uplink 'at the front of the queue'. 
  - We generally lost this item often, as the `AFCntDown` was generally too low while handling the nack anyway.
  - The push is not destructive towards the queue.
- It should no longer be possible for the AS to fail downlink queue operations. Any downlink queue operations that we do now go via `as.attemptDownlinkQueueOp`, which will automatically sync the session between the AS and the NS.

#### Testing

<!-- How did you verify that this change works? -->

I've tested this locally with a 1.0.3A device in class A without any issues. My queue was carried across rejoins (something which was unstable before), and I didn't encounter at any points dropped items.

~Nonetheless this requires better class C testing, as class C was generally the 'catalyst' for these issues. I will pick this up soon, but the work may already be reviewed.~

##### Regressions

<!-- Please indicate features that this change could affect and how that was tested. -->

This PR changes the logic used by the following situations:
- The queue changes on `JoinAccept`
- The queue changes on `DownlinkQueueInvalidated`
- The queue changes on `DownlinkNack`

All of these have been tested in class A, ~but still require class C testing~.

Edit: Class C is done and looks promising.

#### Notes for Reviewers
<!--
NOTE: This section is optional.

Motivate briefly why it is implemented this way, if that deviates from the
implementation proposal in the referenced issues.
- How should your reviewers approach this pull request?
- @mention reviewers with special requests or questions for them
-->

@johanstokking is the main reviewer.

I consider this PR to be close to water tight. But please poke holes if you see any.

#### Checklist
<!-- Make sure that this pull request is complete. -->

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

adriansmares committed:
as: Refactor downlink queue recalculation logic

johanstokking commented on pkg/applicationserver/applicationserver.go:
```diff
@@ -450,30 +449,70 @@ func (as *ApplicationServer) buildSessionsFromError(ctx context.Context, dev *tt
+		if err == nil {
+			return mask, nil
+		}
+		if attempt >= maxDownlinkQueueOperationAttempts || as.skipPayloadCrypto(ctx, link, dev, nil) {
```
First attempt has `attempt == 0`, while `maxDownlinkQueueOperationAttempts` is a counter. So having that value set to `1` results in two attempts.

johanstokking commented on pkg/applicationserver/applicationserver.go:
```diff
@@ -450,30 +449,70 @@ func (as *ApplicationServer) buildSessionsFromError(ctx context.Context, dev *tt
+		return nil, err
 	}
+	mask := make([]string, 0, 2)
+	attempt := 0
```
```suggestion
	attempt := 1
```

johanstokking commented on pkg/applicationserver/applicationserver.go:
```diff
@@ -780,255 +728,65 @@ func (as *ApplicationServer) handleJoinAccept(ctx context.Context, ids ttnpb.End
 			}
+
+			if session := dev.Session; session != nil {
+				session.LastAFCntDown = items[0].FCnt - 1
```
Why is this? I thought that `LastAFCntDown` would be the last used `FCntDown`, i.e. `items[len(items)-1].FCnt`.

BTW, are we sure `FCnt` is never `0`?

adriansmares replied:
I had a long comment here, which I will leave out for historical reasons, but the TLDR is probably that it would be better to just push the uplinks only to the pending session, if possible. I'll check out if I can do it cleanly.

---

There are two mechanisms here, which actually probably should end up as godoc.

1. We call them invalidated downlinks, but the queue itself has not been cleaned for the `dev.Session` -  we just receive a copy. See https://github.com/TheThingsNetwork/lorawan-stack/blob/c0e437afc7fccb9d506e05f73b4a2d2c3db54d13/pkg/networkserver/downlink.go#L1667-L1690

2. `as.AttempDownlinkQueueOp` will replace both queues (as you cannot replace only one queue). As such, it has to compute the FCnt of each uplink again. If we don't lower this `FCnt` before calling it, each time a device joins, the `FCnt` of each element will increase by the queue size. 

Take this example:
- You schedule 3 downlinks to your device, with FCnts 42, 43, 44. `LastAFCntDown` is 44.
- Your device rejoins. We receive these downlinks in `InvalidatedDownlinks`. 
- We have to do a `replace` operation in order to create the queue for the pending session.
- When we calculate the downlinks for `dev.Session`, we start from 44, so we'll have 45, 46, 47.

If your device rejoins too many times at some point we will run out of FCnts. As such, we attempt to keep the `LastAFCntDown` steady.

johanstokking replied:
> * You schedule 3 downlinks to your device, with FCnts 42, 43, 44. `LastAFCntDown` is 44.
> [...]
> * When we calculate the downlinks for `dev.Session`, we start from 44, so we'll have 45, 46, 47.
> 
> If your device rejoins too many times at some point we will run out of FCnts. As such, we attempt to keep the `LastAFCntDown` steady.

Wait, the `LastAFCntDown` survives a session? That doesn't work with rollovers in the frame.

adriansmares replied:
No - the new `dev.PendingSession` starts with `LastAFCntDown=0`, and then we push the old downlink queue to this pending session. So if you had `n` downlinks in `dev.Session`, your `dev.PendingSession` will contain `n` downlinks, with framecounters between `1` and `n`, and `LastAFCntDown=n`.

The behavior that I described where the counter keeps increasing for `dev.Session` is no longer applicable in any shape or form, after 33c6276310b003a1a18f9c63706e27978451e976. Now on `JoinAccept` we don't even touch the downlink queue of `dev.Session`, we just create a new one for `dev.PendingSession`, for which the counter starts at 0.

johanstokking commented on pkg/applicationserver/applicationserver.go:
```diff
@@ -1171,94 +902,93 @@ func (as *ApplicationServer) handleDownlinkQueueInvalidated(ctx context.Context,
+				return nil, nil, err
+			}
+
+			items := []*ttnpb.ApplicationDownlink{msg}
```
Should we already bump the FCnt here?

adriansmares replied:
`attemptDownlinkQueueOp` will bump it automatically.

johanstokking reviewed (commented):
Looks great. Love all the red blocks of code

adriansmares committed:
as: Avoid replacing the downlink queue on join accept

adriansmares committed:
ns: Provide error details for unmatched downlinks

adriansmares commented:
I've refactored the `JoinAccept` procedure to be replace-free, doing a push on the pending session instead. I've also fixed an NS error which didn't return the session diagnostics. This fixes a specific scenario during which downlinks were lost when the device joined after a confirmed downlink was scheduled.

LE: Class C tests are also pretty good. I no longer observe downlinks being dropped spuriously while mixing confirmed and unconfirmed downlinks.
